### PR TITLE
[TRTLLM-10382][feat] Support rejection sampling in one-model spec dec

### PR DESCRIPTION
**Pull request by ziyixiong-nv**
## Summary by CodeRabbit

* **New Features**
  * Added rejection sampling support for draft token selection during speculative decoding with Eagle3 and MTP models.

* **Configuration**
  * New `use_rejection_sampling` option enables rejection sampling-based draft token acceptance in speculative decoding (requires `allow_advanced_sampling`).

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/speculative/interface.py [hunk 1]
```diff
@@ -523,6 +533,104 @@ def _sample_and_accept_draft_tokens_base(
 
         return accepted_tokens, num_accepted_tokens
 
+    def _can_use_rejection_sampling(self, spec_metadata: SpecMetadata,
+                                    num_contexts: int) -> bool:
+        """
+        Whether rejection sampling can be used for the given spec_metadata and number of context requests.
+        """
+        return spec_metadata.use_rejection_sampling and spec_metadata.draft_probs is not None and num_contexts == 0
```
[the remark this answers is not in the document]
Comment (mikeiovine) [reply tensorrt_llm/_torch/speculative/interface.py]: Piggybacking requests should be relatively infrequent - I think it's fine to skip rejection sampling for these batches. We should prioritize whatever is simpler to implement IMO

File: tensorrt_llm/_torch/speculative/one_model_sampler.py [hunk 2]
```diff
@@ -95,3 +95,85 @@ def sampling_batch_spec_dec_one_model(
...
+    return probs
+
+
+def rejection_sampling_one_model(
+    draft_probs: torch.Tensor,
+    draft_token_ids: torch.Tensor,
+    target_probs: torch.Tensor,
+    deterministic: bool = True,
+    seed: Optional[int] = None,
+    offset: Optional[int] = None,
```
[the remark this answers is not in the document]
Comment (ziyixiong-nv) [reply tensorrt_llm/_torch/speculative/one_model_sampler.py]: Thanks. Updated to use device memory.

File: tensorrt_llm/_torch/speculative/interface.py [hunk 3]
```diff
@@ -540,6 +550,135 @@ def _sample_and_accept_draft_tokens_base(
...
+
+        else:
+            full_draft_probs = draft_probs
+
+        full_draft_tokens = draft_tokens.to(torch.int32).contiguous()
+
+        # Lazily initialize seed/offset tensors on correct device
+        if self.seed is None:
+            self.seed = torch.tensor([0], dtype=torch.int64, device=device)
+            self.offset = torch.tensor([0], dtype=torch.int64, device=device)
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/speculative/interface.py]: Is it possible to put this in `__init__`? 

There's one possible issue with this that makes me uneasy. If the initialization gets captured in a CUDA graph, the seed and offset will be the same every iter.

We're safe from that today because we run 1 or 2 warmup runs before the CUDA graph gets captured, so the goal of this change is to eliminate future bug possibilities.

File: tensorrt_llm/_torch/speculative/interface.py [hunk 4]
```diff
@@ -559,6 +698,64 @@ def _draft_sampler_greedy(self, logits: torch.Tensor, d2t=None):
...
+        """
+        Compute draft probabilities from draft logits and store in spec_metadata
+        for use in the next iteration's rejection sampling.
+
+        Args:
+            draft_logits_list: List of [batch_size, vocab_size] tensors, one per draft layer
+            spec_metadata: Speculative decoding metadata
+            batch_size: Number of requests
+        """
+        num_draft_layers = len(draft_logits_list)
```
[the remark this answers is not in the document]
Comment (ziyixiong-nv) [reply tensorrt_llm/_torch/speculative/interface.py]: Updated in the latest version.

File: tensorrt_llm/_torch/speculative/interface.py [hunk 5]
```diff
@@ -559,6 +698,64 @@ def _draft_sampler_greedy(self, logits: torch.Tensor, d2t=None):
...
+            draft_top_ks = None
+            draft_top_ps = None
+
+        # Initialize preallocated buffer once
+        if spec_metadata.draft_probs is None:
+            # Allocate flat buffer: [max_num_requests * max_draft_len * vocab_size]
+            buffer_size = spec_metadata.max_num_requests * spec_metadata.max_draft_len * vocab_size
+            spec_metadata.draft_probs = torch.empty(buffer_size,
+                                                    dtype=torch.float32,
+                                                    device=device)
```
[the remark this answers is not in the document]
Comment (ziyixiong-nv) [reply tensorrt_llm/_torch/speculative/interface.py]: Updated by moving the initialization to the `prepare` function.

**Review by mikeiovine — APPROVED** [review]

**Review by StanleySun639 — APPROVED** [review]

**Review by QiJune — APPROVED** [review]
LGTM

**Comment by ziyixiong-nv** [comment]
It's going to be supported in another PR along with tree-based rejection sampling.
